README.md
# NAME

Weather::YR - Object-oriented interface to Yr.no's weather service.

# VERSION

Version 0.46.

# SYNOPSIS

    use Weather::YR;
    use DateTime::TimeZone;
    use LWP::UserAgent;
    use feature 'say';

    my $yr = Weather::YR->new(
        lang => 'en',
        lat  => 63.5908,
        lon  => 10.7414,
        tz   => DateTime::TimeZone->new( name => 'Europe/Oslo' ),
        ua   => LWP::UserAgent->new(
            agent => 'AcmeWeatherApp/0.9 support@example.com',
        ),
    );

    foreach my $day ( @{$yr->location_forecast->days} ) {
        say 'Weather data for ' . $day->date->ymd . ':';
        say ' ' x 4 . 'Temperature @ ' . $day->temperature->from->hms . ': ' . $day->temperature->celsius;

        foreach my $dp ( @{$day->datapoints} ) {
            say ' ' x 4 . 'Wind direction @ ' . $dp->wind_direction->from->hms . ': ' . $dp->wind_direction->name;
        }
    }

    # If you are interested in the weather right now (*):

    my $now = $yr->location_forecast->now;

    say "It's " . $now->temperature->celsius . "C outside.";
    say "Weather status: " . $now->precipitation->symbol->text;

    # (*) "Right now" is actually lying, as the data from Yr is always
    #     a _forecast_, ie. what the weather will be like. The now()
    #     method simply picks the closest data point in time.

# DESCRIPTION

This is an object-oriented interface to Yr.no's free weather service located at
[https://api.met.no/](https://api.met.no/).

# METHODS

## location\_forecast

Returns a [Weather::YR::LocationForecast](https://metacpan.org/pod/Weather%3A%3AYR%3A%3ALocationForecast) instance.

# TODO

- Improve the documentation.
- Add more tests.
- Add support for more of Yr.no's APIs.

# BUGS

Please report any bugs or feature requests via the web interface at
[https://rt.cpan.org/Public/Dist/Display.html?Name=Weather-YR](https://rt.cpan.org/Public/Dist/Display.html?Name=Weather-YR), or via
the github interface at [https://github.com/toreau/Weather-YR/issues](https://github.com/toreau/Weather-YR/issues).

# AUTHORS

- Tore Aursand, 2014-2021, `toreau@gmail.com`
- Knut-Olav Hoven, 2008-2013, `knut-olav@hoven.ws`

# CONTRIBUTORS

- Andreas Vögele

# LICENSE AND COPYRIGHT

Copyright 2014-2021, Tore Aursand.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

[http://www.perlfoundation.org/artistic\_license\_2\_0](http://www.perlfoundation.org/artistic_license_2_0)

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


Changes
Revision history for Perl extension Weather::YR.

0.44  2021-10-19
      - Added Germany symbol texts. (PR #9, voegelas)
      - Added LWP::Protocol::https as requirement. (PR #8, voegelas)
      - Limit latitude and longitude to four decimals. (PR #7, voegelas)

0.43  2021-10-13
      - Added the English symbol texts again. Not sure where they went after
        they were added in version 0.40. :-/
      - Improved (hopefully) the example code.
      - Misc. minor cleanups/fixes.

0.42  2021-10-13
      - Use Dist::Zilla::Plugin::ReadmeAnyFromPod to generate the README at
        build time.
      - Updated AUTHORS and CONTRIBUTORS section with correct information.

0.41  2021-10-13
      - Updated the README with the correct documentation.

0.40  2021-10-13
      - Added support for English symbol names.
      - Fixed a deep recursion bug.

0.39  2021-10-13
      - Updated to support API v2.0.

0.38  2016-05-27
      - Fixed an error in the documentation.

0.37  2016-05-27
      - Changed to use the new end point URL; http://api.met.no/
      - Replaced XML::Bare with XML::Simple.
      - Misc. code cleanup.

0.36  2015-04-17
      - Fix: Pass on the 'ua' parameter to Weather::Yr::Base when instantiating
        the YR object.

0.35  2015-02-12
      - The 'ua' parameter can now be anything, as long as it is compatible
        with the LWP::UserAgent API.

0.34  2015-01-13
      - Better error-handling when creating datapoints.
      - Removed alpha status; current API is OK to use in production.

0.33  2015-01-08
      - Added XML schema validation for location forecast.

0.32  2014-08-25
      - Improved the documentation.

0.31  2014-08-21
      - Weather::YR now requires Perl version 5.10 or later.
      - Added support for accessing wind directions and wind speeds for any
        given day.
      - Added support for accessing humidity for any given day.
      - Added support for accessing (air) pressure for any given day.
      - Added support for accessing cloudiness for any given day.
      - Added support for accessing fog for any given day.
      - Added support for accessing dew point temperature for any given day.
      - Added support for accessing wind and temperature probability for any
        given data point.

0.30  2014-08-20
      - Module rewritten from scratch for better extensibility.

0.25  2012-01-20
      - Fix: CPAN issue #74112: Misuse of length on array.

0.24  2011-12-18
      - Fix: YR API Location Forecast updated to version 1.8.
      - Fix: YR API Text Forecast updated to version 1.5.

0.23  2010-05-13
      - Fix: Missing dependency declaration on Class::Accessor::Fast.
      - Fix: Configuration overrides were leaking when multiple forecast modules are used.

0.22  2008-05-29
      - Fix: Added prerequisites to Makefile.PL.

0.21  2008-05-26
      - Fix: Wrong package names in SEE ALSO.
      - Fix: Missing SYNOPSIS of central packages.

0.20  2008-05-24
      - Rebranded from YR to Weather::YR.

0.11  2008-04-30
      - Added LICENSE file for GPL2.
      - Updated README file.
      - Added YR library version to web client user agent.

0.10  2008-04-30
      - Support for web service locationforecast.
      - Support for web service textforecast.

0.01  2008-03-17
      - Original version.


lib/Weather/YR.pm
package Weather::YR;
use Moose;
use namespace::autoclean;

extends 'Weather::YR::Base';

use Weather::YR::LocationForecast;
use Weather::YR::TextLocation;

=encoding utf8

=head1 NAME

Weather::YR - Object-oriented interface to Yr.no's weather service.

=head1 VERSION

Version 0.46.

=cut

our $VERSION = '0.46';

=head1 SYNOPSIS

    use Weather::YR;
    use DateTime::TimeZone;
    use LWP::UserAgent;
    use feature 'say';

    my $yr = Weather::YR->new(
        lang => 'en',
        lat  => 63.5908,
        lon  => 10.7414,
        tz   => DateTime::TimeZone->new( name => 'Europe/Oslo' ),
        ua   => LWP::UserAgent->new(
            agent => 'AcmeWeatherApp/0.9 support@example.com',
        ),
    );

    foreach my $day ( @{$yr->location_forecast->days} ) {
        say 'Weather data for ' . $day->date->ymd . ':';
        say ' ' x 4 . 'Temperature @ ' . $day->temperature->from->hms . ': ' . $day->temperature->celsius;

        foreach my $dp ( @{$day->datapoints} ) {
            say ' ' x 4 . 'Wind direction @ ' . $dp->wind_direction->from->hms . ': ' . $dp->wind_direction->name;
        }
    }

    # If you are interested in the weather right now (*):

    my $now = $yr->location_forecast->now;

    say "It's " . $now->temperature->celsius . "C outside.";
    say "Weather status: " . $now->precipitation->symbol->text;

    # (*) "Right now" is actually lying, as the data from Yr is always
    #     a _forecast_, ie. what the weather will be like. The now()
    #     method simply picks the closest data point in time.

=head1 DESCRIPTION

This is an object-oriented interface to Yr.no's free weather service located at
L<https://api.met.no/>.

=cut

has 'location_forecast' => ( isa => 'Weather::YR::LocationForecast', is => 'ro', lazy_build => 1 );
# has 'text_location'     => ( isa => 'Weather::YR::TextLocation',     is => 'ro', lazy_build => 1 );

=head1 METHODS

=head2 location_forecast

Returns a L<Weather::YR::LocationForecast> instance.

=cut

sub _build_location_forecast {
    my $self = shift;

    return Weather::YR::LocationForecast->new(
        lat  => $self->lat,
        lon  => $self->lon,
        msl  => $self->msl,
        xml  => $self->xml,
        lang => $self->lang,
        tz   => $self->tz,
        ua   => $self->ua,
    );
}

# sub _build_text_location {
#     my $self = shift;

#     return Weather::YR::TextLocation->new(
#         lat  => $self->lat,
#         lon  => $self->lon,
#         lang => $self->lang,
#     );
# }

__PACKAGE__->meta->make_immutable;

1;

=head1 TODO

=over 4

=item * Improve the documentation.

=item * Add more tests.

=item * Add support for more of Yr.no's APIs.

=back

=head1 BUGS

Please report any bugs or feature requests via the web interface at
L<https://rt.cpan.org/Public/Dist/Display.html?Name=Weather-YR>, or via
the github interface at L<https://github.com/toreau/Weather-YR/issues>.

=head1 AUTHORS

=over 4

=item * Tore Aursand, 2014-2021, C<< toreau@gmail.com >>

=item * Knut-Olav Hoven, 2008-2013, C<< knut-olav@hoven.ws >>

=back

=head1 CONTRIBUTORS

=over 4

=item * Andreas Vögele

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2014-2021, Tore Aursand.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
